src/Config.h

```diff
@@ -0,0 +1,1488 @@
+            return _doc["hardware"]["sensors"]["watertank"]["mode"];
+        }
+
+        void setWaterTankSensorMode(const int value) {
+            _doc["hardware"]["sensors"]["watertank"]["mode"] = value;
+        }
+
+        // Scale
+        bool getScaleEnabled() {
+            return _doc["hardware"]["sensors"]["scale"]["enabled"] | false;
+        }
+
+        void setScaleEnabled(const bool value) {
+            _doc["hardware"]["sensors"]["scale"]["enabled"] = value;
+        }
+
+        int getScaleSamples() {
+            return _doc["hardware"]["sensors"]["scale"]["samples"];
+        }
+
+        void setScaleSamples(const int value) {
+            _doc["hardware"]["sensors"]["scale"]["samples"] = value;
+        }
+
+        int getScaleType() {
+            return _doc["hardware"]["sensors"]["scale"]["type"];
+        }
+
+        void setScaleType(const int value) {
+            _doc["hardware"]["sensors"]["scale"]["type"] = value;
+        }
+
+        float getScaleCalibration() {
+            return _doc["hardware"]["sensors"]["scale"]["calibration"] | static_cast<float>(SCALE_CALIBRATION_FACTOR);
+        }
+
+        void setScaleCalibration(const float value) {
+            _doc["hardware"]["sensors"]["scale"]["calibration"] = constrain(value, -100000.0f, 100000.0f);
+        }
+
+        void setScaleKnownWeight(const float value) {
+            _doc["hardware"]["sensors"]["scale"]["known_weight"] = constrain(value, 0.0f, 2000.0f);
+        }
+
+        float getScaleKnownWeight() {
+            return _doc["hardware"]["sensors"]["scale"]["known_weight"] | static_cast<float>(SCALE_KNOWN_WEIGHT);
+        }
+
+        float getScale2Calibration() {
+            return _doc["hardware"]["sensors"]["scale"]["calibration2"] | static_cast<float>(SCALE_CALIBRATION_FACTOR);
+        }
+
+        void setScale2Calibration(const float value) {
+            _doc["hardware"]["sensors"]["scale"]["calibration2"] = constrain(value, -100000.0f, 100000.0f);
+        }
+
+    private:
+        inline static auto CONFIG_FILE = "/config.json";
+
+        StaticJsonDocument<4096> _doc;
+
+        bool validateAndApplyConfig(const JsonDocument& doc) {
+            // PID parameters
+            if (doc["pid"].containsKey("enabled")) {
+                if (!doc["pid"]["enabled"].is<bool>()) {
+                    return false;
+                }
+
+                setPidEnabled(doc["pid"]["enabled"].as<bool>());
+            }
+
+            if (doc["pid"].containsKey("use_ponm")) {
+                if (!doc["pid"]["use_ponm"].is<bool>()) {
+                    return false;
+                }
+
+                setUsePonM(doc["pid"]["use_ponm"].as<bool>());
+            }
+
+            if (doc["pid"].containsKey("ema_factor")) {
+                const double value = doc["pid"]["ema_factor"].as<double>();
+
+                if (!validateParameterRange("pid.ema_factor", value, PID_EMA_FACTOR_MIN, PID_EMA_FACTOR_MAX)) {
```

Comment:
Example here, you use `constrain(/* ... */)` above and here again validate the constrained range - I think this is a bit duplicated